specs/agents/metrics-otel.md

```diff
@@ -0,0 +1,181 @@
+
+However, agents MUST support explicit bucket histogram aggregations with custom boundaries provided by users via views.
+
+Before export, the OpenTelemetry histogram bucket data needs to be converted to IntakeV2 histogram data. Because the OpenTelemetry specification allows negative values in histograms, the data needs to be converted to T-Digests: For T-Digests, the IntakeV2 metricset `samples.values` properties holds the **midpoints** of the histogram buckets instead of the upper/lower bounds.
+
+Therefore, the following algorithm needs to be used to convert [OpenTelemetry explicit bucket histogram data](https://opentelemetry.io/docs/reference/specification/metrics/data-model/#histogram) to IntakeV2 histograms:
+
+```
+IntakeV2Histo convertBucketBoundaries(OtelHisto input) {
+    List<Double> intakeCounts = new ArrayList<>();
+    List<Double> intakeValues = new ArrayList<>();
+
+    List<Double> otelCounts = input.getCounts();
+    List<Double> otelBoundaries = input.getBoundaries();
+    
+    int bucketCount = otelCounts.size();
+    // otelBoundaries has a size of bucketCount-1
+    // the first bucket has the boundaries ( -inf, otelBoundaries[0] ]
+    // the second bucket has the boundaries ( otelBoundaries[0], otelBoundaries[1] ]
+    // ..
+    // the last bucket has the boundaries (otelBoundaries[bucketCount-2], inf)
+    
+    for (int i = 0; i < bucketCount; i++) {
+        if (otelCounts.get(i) != 0) { //ignore empty buckets
+            intakeCounts.add(otelCounts.get(i));
+            if (i == 0) { //first bucket
+                double bounds = otelBoundaries.get(0);
+                if(bounds > 0) {
+                    bounds /= 2;
+                }
+                intakeValues.add(bounds);
+            } else if (i == bucketCount - 1) { //last bucket
+                intakeValues.add(otelBoundaries.get(bucketCount - 2));
+            } else { //in between
+                double lower = otelBoundaries.get(i - 1);
+                double upper = otelBoundaries.get(i);
+                intakeValues.add(lower + (upper - lower) / 2);
+            }
+        }
+    }
+    
+    return new IntakeV2Histo(intakeCounts, intakeValues);
+}
+```
+
+The same algorithm is used by the APM server to convert OTLP histograms.
+The `sum`, `count`, `min` and `max` within the OpenTelemetry histogram data are discarded for now until we support them in IntakeV2 histograms.
+
+## Labels
+
+Conceptually, elastic metric labels keys correspond to [OpenTelemetry Attributes](https://opentelemetry.io/docs/reference/specification/common/#attribute).
+When exporting, the agents MUST convert the attributes to labels as described in this section and MUST group metrics with the same attributes and OpenTelemetry instrumentation scope together into a single metricset.
```

Comment:
For the OTel tracing bridge, we introduced an `otel.attributes` property in the intake v2 spec: https://github.com/elastic/apm/blob/main/specs/agents/tracing-api-otel.md#attributes-mapping.

APM Server then maps known attributes to ECS and unknown attributes to labels. I wonder if we should do the same for metrics. Does APM Server currently map OTel metric attributes to ECS fields?

Maybe this isn't as relevant to metrics as it is to traces if the metric attributes are mostly used for custom user-defined attributes.

Reply:
The server maps the [metadata](https://github.com/elastic/apm-data/blob/main/input/otlp/metrics.go#L75) and some known and [defined metrics](https://github.com/elastic/apm-data/blob/main/input/otlp/metrics.go#L81).